unitary/alpha/qudit_state_transform.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2022 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import itertools
+
+import numpy as np
+
+
+def _nearest_power_of_two_ceiling(qudit_dim: int) -> int:
+    """Returns the smallest power of two greater than or equal to qudit_dim."""
+    if qudit_dim == 0:
+        return 0
+    result = 1
+    while result < qudit_dim:
+        result = result << 1
+    return result
+
+
+def qudit_to_qubit_state(
+    qudit_dimension: int,
+    num_qudits: int,
+    qudit_state_vector: np.ndarray,
+    pad_value: np.complex_ = 0,
+) -> np.ndarray:
+    """Converts a qudit-space quantum state vector to m-qubit-per-qudit column vector.
+
+    Each qudit is replaced by a set of qubits. Since the set of qubits can represent a larger state
+    space than the qudit, the state vector needs to be padded with 0s for those extra elements.
+    Based on https://drive.google.com/file/d/1n1Ym7JdnM44NnvNQDUXSk5rBLTQZzJ9t and
+    https://colab.research.google.com/drive/1MC6D4FyOXG0RjvzyV9IFcdRsLbUwwAcx
+
+    Args:
+        qudit_dimension: The dimension of a single qudit i.e. the number of states it can
+          represent.
+        num_qudits: The number of qudits in the given state vector.
+        qudit_state_vector: A flat, 1-D numpy array representing the state vector in qudit form.
```

Comment:
It might be worth stating the expected size of this vector, just for clarity's sake.

Reply:
Done